pkg/workspace/resource/node_claim.go

```diff
@@ -0,0 +1,380 @@
+	"github.com/kaito-project/kaito/pkg/utils/nodeclaim"
+	"github.com/kaito-project/kaito/pkg/utils/plugin"
+	"github.com/kaito-project/kaito/pkg/utils/status"
+)
+
+type NodeClaimManager struct {
+	client.Client
+	recorder     record.EventRecorder
+	expectations *utils.ControllerExpectations
+	logger       klog.Logger
+}
+
+func NewNodeClaimManager(c client.Client, recorder record.EventRecorder, expectations *utils.ControllerExpectations) *NodeClaimManager {
+	return &NodeClaimManager{
+		Client:       c,
+		recorder:     recorder,
+		expectations: expectations,
+		logger:       klog.NewKlogr().WithName("NodeClaim"),
+	}
+}
+
+func (c *NodeClaimManager) SyncNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace) (bool, error) {
+	// ensure nodeclaims for workspace
+	if err := c.ensureNodeClaims(ctx, wObj); err != nil {
+		return false, err
+	}
+
+	// check all nodeclaims are ready or not
+	return c.areNodeClaimsReady(ctx, wObj)
+}
+
+// ensureNodeClaims ensures the correct number of NodeClaims for the workspace
+// based on the TargetNodeCount in the workspace status, considering preferred nodes
+func (c *NodeClaimManager) ensureNodeClaims(ctx context.Context, wObj *kaitov1beta1.Workspace) error {
+	workspaceKey := client.ObjectKeyFromObject(wObj).String()
+
+	// Check if we should wait for expectations to be satisfied
+	if !c.expectations.SatisfiedExpectations(c.logger, workspaceKey) {
+		klog.V(4).InfoS("Waiting for NodeClaim expectations to be satisfied",
+			"workspace", klog.KObj(wObj))
+		return nil
+	}
+
+	// Calculate the number of NodeClaims needed (target - preferred nodes)
+	requiredNodeClaimsCount, err := nodeclaim.GetRequiredNodeClaimsCount(ctx, c.Client, wObj)
+	if err != nil {
+		return fmt.Errorf("failed to get required NodeClaims: %w", err)
+	}
+
+	klog.V(4).InfoS("NodeClaim calculation",
+		"workspace", klog.KObj(wObj),
+		"requiredNodeClaims", requiredNodeClaimsCount)
+
+	// Get existing NodeClaims for this workspace
+	existingNodeClaims, err := nodeclaim.GetExistingNodeClaims(ctx, c.Client, wObj)
+	if err != nil {
+		return fmt.Errorf("failed to get existing NodeClaims: %w", err)
+	}
+
+	currentNodeClaimCount := len(existingNodeClaims)
+
+	if currentNodeClaimCount < requiredNodeClaimsCount {
+		// Need to create more NodeClaims
+		nodesToCreate := requiredNodeClaimsCount - currentNodeClaimCount
+		klog.InfoS("Creating additional NodeClaims",
+			"workspace", klog.KObj(wObj),
+			"current", currentNodeClaimCount,
+			"required", requiredNodeClaimsCount,
+			"toCreate", nodesToCreate)
+
+		// Update status condition to indicate NodeClaims are being created
+		if updateErr := status.UpdateStatusConditionIfNotMatch(ctx, c.Client, wObj, kaitov1beta1.ConditionTypeNodeClaimStatus, metav1.ConditionFalse,
+			"CreatingNodeClaims", fmt.Sprintf("Creating %d additional NodeClaims (current: %d, required: %d)", nodesToCreate, currentNodeClaimCount, requiredNodeClaimsCount)); updateErr != nil {
```

Comment:
This condition is only set to true once. I can't find where this condition is being set to False once creation is complete. Do we need this condition at all?